### PR TITLE
[BEAM-2899] Add A FnService, FnServer to runner Fn Execution

### PR DESCRIPTION
**Pull request by tgroh**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
A FnService is a closeable BindableService. A FnServer is exactly one
FnService and a server which exposes it as an endpoint.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @lukecwik

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/GrpcFnServer.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.fnexecution;
+
+import io.grpc.Server;
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+import org.apache.beam.model.pipeline.v1.Endpoints;
```
Comment (lukecwik) [inline runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/GrpcFnServer.java]: nit: Its strange to import both Endpoints and Endpoints.ApiServiceDescriptor, please choose one

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/GrpcFnServer.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  public Endpoints.ApiServiceDescriptor getApiServiceDescriptor() {
+    return apiServiceDescriptor;
+  }
+
+  /** Get the service exposed by this {@link GrpcFnServer}. */
+  public ServiceT getService() {
+    return service;
+  }
+
+  public void awaitTermination() throws InterruptedException {
```
Comment (lukecwik) [inline runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/GrpcFnServer.java]: Is there a reason to expose this, I would expect that we would only ever expose close()?

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/GrpcFnServer.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+  @Override
+  public void close() throws Exception {
+    try {
+      // The server has been closed, and should not respond to any new incoming calls.
+      server.shutdown();
+      service.close();
+      server.awaitTermination(60, TimeUnit.SECONDS);
+    } finally {
+      server.shutdownNow();
```
Comment (lukecwik) [inline runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/GrpcFnServer.java]: You'll want another awaitTermination after server.shutdownNow.

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/GrpcFnServer.java (hunk 1 above)
Comment (tgroh) [reply runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/GrpcFnServer.java]: Done.

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/GrpcFnServer.java (hunk 2 above)
Comment (tgroh) [reply runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/GrpcFnServer.java]: Removed.

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/GrpcFnServer.java (hunk 3 above)
Comment (tgroh) [reply runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/GrpcFnServer.java]: Done.

**Comment by lukecwik** [comment]
LGTM

**Pull request closed without merging**